metricbeat/module/iis/application_pool/reader.go

```diff
@@ -0,0 +1,276 @@
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// +build windows
+
+package application_pool
+
+import (
+	"strings"
+
+	"github.com/elastic/go-sysinfo"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/helper/windows/pdh"
+	"github.com/elastic/beats/metricbeat/mb"
+)
+
+// Reader  strucr will contain the pdh query and config options
+type Reader struct {
+	Query            pdh.Query         // PDH Query
+	ApplicationPools []ApplicationPool // Mapping of counter path to key used for the label (e.g. processor.name)
+	log              *logp.Logger      // logger
+	hasRun           bool              // will check if the reader has run a first time
+	WorkerProcesses  map[string]string
+}
+
+// ApplicationPool struct contains the list of applications and their worker processes
+type ApplicationPool struct {
+	Name             string
+	WorkerProcessIds []int
+	counters         map[string]string
+}
+
+// WorkerProcess struct contains the worker process details
+type WorkerProcess struct {
+	ProcessId    int
+	InstanceName string
+}
+
+const ecsProcessId = "process.pid"
+
+var appPoolCounters = map[string]string{
+	"process.pid":                          "\\Process(w3wp*)\\ID Process",
+	"process.cpu_usage_perc":               "\\Process(w3wp*)\\% Processor Time",
+	"process.handle_count":                 "\\Process(w3wp*)\\Handle Count",
+	"process.thread_count":                 "\\Process(w3wp*)\\Thread Count",
+	"process.working_set":                  "\\Process(w3wp*)\\Working Set",
+	"process.private_byte":                 "\\Process(w3wp*)\\Private Bytes",
+	"process.virtual_bytes":                "\\Process(w3wp*)\\Virtual Bytes",
+	"process.page_faults_per_sec":          "\\Process(w3wp*)\\Page Faults/sec",
+	"process.io_read_operations_per_sec":   "\\Process(w3wp*)\\IO Read Operations/sec",
+	"process.io_write_operations_per_sec":  "\\Process(w3wp*)\\IO Write Operations/sec",
+	"net_clr.total_exceptions_thrown":      "\\.NET CLR Exceptions(w3wp*)\\# of Exceps Thrown",
+	"net_clr.exceptions_thrown_per_sec":    "\\.NET CLR Exceptions(w3wp*)\\# of Exceps Thrown / sec",
+	"net_clr.filters_per_sec":              "\\.NET CLR Exceptions(w3wp*)\\# of Filters / sec",
+	"net_clr.finallys_per_sec":             "\\.NET CLR Exceptions(w3wp*)\\# of Finallys / sec",
+	"net_clr.throw_to_catch_depth_per_sec": "\\.NET CLR Exceptions(w3wp*)\\Throw To Catch Depth / sec",
+}
```

Comment:
We should also add .NET specific metrics:

* `ASP.NET Applications(__Total__)\Cache Total Trims`
Total number of entries forcibly removed from
the cache due to memory pressure.
* `ASP.NET Applications(__Total__)\Cache Total Entries`
Total number of entries within the cache (both
internal and user added)
* `ASP.NET Applications(__Total__)\Cache Total Hit Ratio`
Ratio of hits from all cache calls.
* `ASP.NET Applications(__Total__)\Cache Total Turnover Rate`
Number of additions and removals to the total
cache per second.
* `ASP.NET Applications(__Total__)\Output Cache Entries`
Current number of entries in the output cache.
* `ASP.NET Applications(__Total__)\Output Cache Hits`
Total number of output cacheable requests served from the output cache.
* `ASP.NET Applications(__Total__)\Output Cache Hit Ratio`
Ratio of hits to requests for output cacheable requests.
* `ASP.NET Applications(__Total__)\Output Cache Turnover Rate`
Number of additions and removals to the output cache per second.
* `ASP.NET Applications(__Total__)\Compilations Total`
Number of .asax, .ascx, .ashx, .asmx, or .aspx source files dynamically compiled.
* `ASP.NET Applications(__Total__)\Errors Total/Sec`
Rate of errors occurred.
* `ASP.NET Applications(__Total__)\Pipeline Instance Count`
Number of active pipeline instances.
* `ASP.NET Applications(__Total__)\Requests Executing`
The number of requests currently executing.
* `ASP.NET Applications(__Total__)\Requests in Application Queue`
The number of requests in the application request queue.
* `ASP.NET Applications(__Total__)\Requests/Sec` 
The number of requests executed per second.
* `ASP.NET\Application Restarts` 
Number of times the application has been restarted during the Web server's lifetime.
* `ASP.NET\Request Wait Time` 
The number of milliseconds the most recent request was waiting in the queue.

Reply:
skip all except
```
ASP.NET Applications(__Total__)\Errors Total/Sec
ASP.NET Applications(__Total__)\Pipeline Instance Count
ASP.NET Applications(__Total__)\Requests Executing
ASP.NET Applications(__Total__)\Requests in Application Queue
ASP.NET Applications(__Total__)\Requests/Sec
ASP.NET\Application Restarts
ASP.NET\Request Wait Time
```